clustered/server/src/main/java/org/ehcache/clustered/server/offheap/OffHeapChainMap.java

```diff
@@ -0,0 +1,255 @@
+ * limitations under the License.
+ */
+package org.ehcache.clustered.server.offheap;
+
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.concurrent.locks.Lock;
+
+import org.ehcache.clustered.common.store.Chain;
+import org.ehcache.clustered.common.store.Element;
+import org.ehcache.clustered.common.store.Util;
+import org.terracotta.offheapstore.MapInternals;
+
+import org.terracotta.offheapstore.ReadWriteLockedOffHeapClockCache;
+import org.terracotta.offheapstore.paging.PageSource;
+import org.terracotta.offheapstore.storage.portability.Portability;
+
+public class OffHeapChainMap<K> implements MapInternals {
+
+  private final ReadWriteLockedOffHeapClockCache<K, InternalChain> heads;
+  private final OffHeapChainStorageEngine chainStorage;
+
+  public OffHeapChainMap(PageSource source, Portability<? super K> keyPortability) {
+    this.chainStorage = new OffHeapChainStorageEngine(source, keyPortability);
+    this.heads = new ReadWriteLockedOffHeapClockCache<K, InternalChain>(source, chainStorage);
+  }
+
+  public Chain get(K key) {
+    final Lock lock = heads.readLock();
+    lock.lock();
+    try {
+      InternalChain chain = heads.get(key);
+      if (chain == null) {
+        return EMPTY_CHAIN;
+      } else {
+        try {
+          return chain.detach();
+        } finally {
+          chain.close();
+        }
+      }
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  public Chain getAndAppend(K key, ByteBuffer element) {
+    final Lock lock = heads.writeLock();
+    lock.lock();
+    try {
+      InternalChain chain = heads.get(key);
+      if (chain == null) {
+        heads.put(key, chainStorage.newChain(element));
+        return EMPTY_CHAIN;
+      } else {
+        try {
+          Chain current = chain.detach();
+          if (!chain.append(element)) {
+            heads.remove(key).close();
+          }
+          return current;
+        } finally {
+          chain.close();
+        }
+      }
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  public void append(K key, ByteBuffer element) {
+    final Lock lock = heads.writeLock();
+    lock.lock();
+    try {
+      InternalChain chain = heads.get(key);
+      if (chain == null) {
+        heads.put(key, chainStorage.newChain(element));
+      } else {
+        try {
+          if (!chain.append(element)) {
+            heads.remove(key).close();
+          }
+        } finally {
+          chain.close();
+        }
+      }
+    } finally {
+      lock.unlock();
+    }
+
+  }
+
+  public boolean replaceAtHead(K key, Chain expected, Chain replacement) {
+    final Lock lock = heads.writeLock();
+    lock.lock();
+    try {
+      InternalChain chain = heads.get(key);
+      if (chain == null) {
+        if (expected.iterator().hasNext()) {
+          return false;
+        } else {
```

Comment:
Fail to see the use case for this. Isn't is simpler to throw `IllegalArgumentException` when receiving an empty `expected`?